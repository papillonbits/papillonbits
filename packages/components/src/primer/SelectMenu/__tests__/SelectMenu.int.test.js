import { ariaAttr, items, custom, regular } from './SelectMenu.int.story'

jest.mock('@papillonbits/library/array', () => ({
  getIndexItems: () => {},
  getIndexItemsWithSelected: () => {},
}))

jest.mock('@papillonbits/library/hooks', () => ({
  useRef: () => {},
  useState: () => {},
}))

const libraryArrayMockObject = require('@papillonbits/library/array')
const libraryHooksMockObject = require('@papillonbits/library/hooks')

describe('<SelectMenu />', () => {
  let indexItemsDataObject = [
    { ...items[0], index: 0 },
    { ...items[1], index: 1 },
    { ...items[2], index: 2 },
    { ...items[3], index: 3 },
  ]

  const newIndexItemsWithSelectedDataObject = [
    { ...items[0], index: 0, isSelected: true },
    { ...items[1], index: 1, isSelected: false },
    { ...items[2], index: 2, isSelected: false },
    { ...items[3], index: 3, isSelected: false },
  ]

  const newItemsDataObject = [
    { ...items[0], isSelected: true },
    { ...items[1], isSelected: false },
    { ...items[2], isSelected: false },
    { ...items[3], isSelected: false },
  ]

  const onClickMockFn = jest.fn()

  const setIndexItemsMockFn = jest.fn((indexItems) => {
    indexItemsDataObject = indexItems
  })

  beforeEach(() => {
    jest.spyOn(libraryArrayMockObject, 'getIndexItems').mockReturnValue(indexItemsDataObject)

    jest.spyOn(libraryArrayMockObject, 'getIndexItemsWithSelected').mockReturnValue(newIndexItemsWithSelectedDataObject)

    jest.spyOn(libraryHooksMockObject, 'useState').mockImplementation(() => [indexItemsDataObject, setIndexItemsMockFn])

    jest.spyOn(libraryHooksMockObject, 'useRef').mockImplementation(() => ({
      current: {
        removeAttribute: () => {},
      },
    }))
  })

  afterEach(() => jest.clearAllMocks())

  describe('Event', () => {
    test('must return new items without index when onClick()', () => {
      const mountCustomRender = global.renderMount(custom(onClickMockFn))
      mountCustomRender.find('button').last().simulate('click')
      expect(mountCustomRender.props().onClick).toBe(onClickMockFn)
      expect(mountCustomRender.props().onClick).toHaveBeenCalledWith({
        ariaAttr,
        items: newItemsDataObject,
      })
    })
  })

  describe('Render', () => {
    test('must match regular()', () => {
      expect(global.renderToJSON(regular())).toMatchSnapshot()
    })
  })

  describe('State', () => {
    test('must reset new items with index when onClick()', () => {
      const mountCustomRender = global.renderMount(custom(onClickMockFn))
      mountCustomRender.find('button').last().simulate('click')
      expect(setIndexItemsMockFn).toHaveBeenCalledWith(newIndexItemsWithSelectedDataObject)
      expect(indexItemsDataObject).toBe(newIndexItemsWithSelectedDataObject)
    })
  })
})
