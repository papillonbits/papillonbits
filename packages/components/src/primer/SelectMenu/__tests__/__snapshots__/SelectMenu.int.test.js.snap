// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectMenu /> Render must match regular() 1`] = `
<details
  className="details-reset details-overlay"
>
  <summary
    aria-haspopup={true}
    className="btn btn-outline"
  >
    Choose an item
     
    <mock-file
      aria-hidden="true"
      aria-label={null}
      className="ml-2"
      height={16}
      role="img"
      style={
        Object {
          "display": "inline-block",
          "fill": "currentColor",
          "userSelect": "none",
          "verticalAlign": "text-bottom",
        }
      }
      viewBox="0 0 16 16"
      width={16}
    />
  </summary>
  <div
    className="SelectMenu"
  >
    <div
      className="SelectMenu-modal"
    >
      <header
        className="SelectMenu-header"
      >
        <h3
          className="SelectMenu-title"
        >
          Title
        </h3>
        <button
          className="SelectMenu-closeButton"
          type="button"
        />
      </header>
      <div
        className="SelectMenu-list"
      >
        <button
          className="SelectMenu-item"
          onClick={[Function]}
          role="menuitem"
          type="button"
        >
          Item 1
        </button>
        <button
          className="SelectMenu-item"
          onClick={[Function]}
          role="menuitem"
          type="button"
        >
          Item 2
        </button>
        <button
          className="SelectMenu-item"
          onClick={[Function]}
          role="menuitem"
          type="button"
        >
          Item 3
        </button>
        <button
          className="SelectMenu-item"
          onClick={[Function]}
          role="menuitem"
          type="button"
        >
          Item 4
        </button>
      </div>
    </div>
  </div>
</details>
`;
